File: api/AltV.Net/Elements/Entities/Player.cs
using System;
using System.Runtime.InteropServices;
using AltV.Net.Data;
using AltV.Net.Elements.Args;
using AltV.Net.Native;

namespace AltV.Net.Elements.Entities
{
    public class Player : Entity, IPlayer
    {
        public static ushort GetId(IntPtr playerPointer) => AltNative.Player.Player_GetID(playerPointer);

        public override uint Model
        {
            get
            {
                CheckIfEntityExists();
                return AltNative.Player.Player_GetModel(NativePointer);
            }
            set
            {
                CheckIfEntityExists();
                AltNative.Player.Player_SetModel(NativePointer, value);
            }
        }

        public override Position Position
        {
            get
            {
                CheckIfEntityExists();
                var position = Position.Zero;
                AltNative.Player.Player_GetPosition(NativePointer, ref position);
                return position;
            }
            set
            {
                CheckIfEntityExists();
                AltNative.Player.Player_SetPosition(NativePointer, value);
            }
        }

        public override Rotation Rotation
        {
            get
            {
                CheckIfEntityExists();
                var rotation = Rotation.Zero;
                AltNative.Player.Player_GetRotation(NativePointer, ref rotation);
                return rotation;
            }
            set
            {
                CheckIfEntityExists();
                AltNative.Player.Player_SetRotation(NativePointer, value);
            }
        }

        public override short Dimension
        {
            get
            {
                CheckIfEntityExists();
                return AltNative.Player.Player_GetDimension(NativePointer);
            }
            set
            {
                CheckIfEntityExists();
                AltNative.Player.Player_SetDimension(NativePointer, value);
            }
        }

        public uint Ping
        {
            get
            {
                CheckIfEntityExists();
                return AltNative.Player.Player_GetPing(NativePointer);
            }
        }

        public override void GetMetaData(string key, ref MValue value) =>
            AltNative.Player.Player_GetMetaData(NativePointer, key, ref value);

        public override void SetMetaData(string key, ref MValue value) =>
            AltNative.Player.Player_SetMetaData(NativePointer, key, ref value);

        public override void SetSyncedMetaData(string key, ref MValue value) =>
            AltNative.Player.Player_SetSyncedMetaData(NativePointer, key, ref value);

        public override void GetSyncedMetaData(string key, ref MValue value) =>
            AltNative.Player.Player_GetSyncedMetaData(NativePointer, key, ref value);

        public bool IsConnected
        {
            get
            {
                CheckIfEntityExists();
                return AltNative.Player.Player_IsConnected(NativePointer);
            }
        }

        public string Name
        {
            get
            {
                CheckIfEntityExists();
                var ptr = IntPtr.Zero;
                AltNative.Player.Player_GetName(NativePointer, ref ptr);
                return Marshal.PtrToStringUTF8(ptr);
            }
            set
            {
                CheckIfEntityExists();
                var stringPtr = AltNative.StringUtils.StringToHGlobalUtf8(value);
                AltNative.Player.Player_SetName(NativePointer, stringPtr);
                Marshal.FreeHGlobal(stringPtr);
            }
        }

        public ushort Health
        {
            get
            {
                CheckIfEntityExists();
                return AltNative.Player.Player_GetHealth(NativePointer);
            }
            set
            {
                CheckIfEntityExists();
                AltNative.Player.Player_SetHealth(NativePointer, value);
            }
        }

        public bool IsDead
        {
            get
            {
                CheckIfEntityExists();
                return AltNative.Player.Player_IsDead(NativePointer);
            }
        }

        public bool IsJumping
        {
            get
            {
                CheckIfEntityExists();
                return AltNative.Player.Player_IsJumping(NativePointer);
            }
        }

        public bool IsInRagdoll
        {
            get
            {
                CheckIfEntityExists();
                return AltNative.Player.Player_IsInRagdoll(NativePointer);
            }
        }

        public bool IsAiming
        {
            get
            {
                CheckIfEntityExists();
                return AltNative.Player.Player_IsAiming(NativePointer);
            }
        }

        public bool IsShooting
        {
            get
            {
                CheckIfEntityExists();
                return AltNative.Player.Player_IsShooting(NativePointer);
            }
        }

        public bool IsReloading
        {
            get
            {
                CheckIfEntityExists();
                return AltNative.Player.Player_IsReloading(NativePointer);
            }
        }

        public ushort Armor
        {
            get
            {
                CheckIfEntityExists();
                return AltNative.Player.Player_GetArmor(NativePointer);
            }
            set
            {
                CheckIfEntityExists();
                AltNative.Player.Player_SetArmor(NativePointer, value);
            }
        }

        public float MoveSpeed
        {
            get
            {
                CheckIfEntityExists();
                return AltNative.Player.Player_GetMoveSpeed(NativePointer);
            }
        }

        public uint Weapon
        {
            get
            {
                CheckIfEntityExists();
                return AltNative.Player.Player_GetWeapon(NativePointer);
            }
        }

        public ushort Ammo
        {
            get
            {
                CheckIfEntityExists();
                return AltNative.Player.Player_GetAmmo(NativePointer);
            }
        }

        public Position AimPosition
        {
            get
            {
                CheckIfEntityExists();
                var position = Position.Zero;
                AltNative.Player.Player_GetAimPos(NativePointer, ref position);
                return position;
            }
        }

        public Rotation HeadRotation
        {
            get
            {
                CheckIfEntityExists();
                var rotation = Rotation.Zero;
                AltNative.Player.Player_GetHeadRotation(NativePointer, ref rotation);
                return rotation;
            }
        }

        public bool IsInVehicle
        {
            get
            {
                CheckIfEntityExists();
                return AltNative.Player.Player_IsInVehicle(NativePointer);
            }
        }

        public IVehicle Vehicle
        {
            get
            {
                CheckIfEntityExists();
                var entityPointer = AltNative.Player.Player_GetVehicle(NativePointer);
                if (entityPointer == IntPtr.Zero) return null;
                return Alt.Module.VehiclePool.GetOrCreate(entityPointer, out var vehicle) ? vehicle : null;
            }
        }

        public byte Seat
        {
            get
            {
                CheckIfEntityExists();
                return AltNative.Player.Player_GetSeat(NativePointer);
            }
        }

        public Player(IntPtr nativePointer, ushort id) : base(nativePointer, BaseObjectType.Player, id)
        {
        }

        public void Spawn(Position position, uint delayMs)
        {
            CheckIfEntityExists();
            AltNative.Player.Player_Spawn(NativePointer, position, delayMs);
        }

        public void Despawn()
        {
            CheckIfEntityExists();
            AltNative.Player.Player_Despawn(NativePointer);
        }

        public void SetDateTime(int day, int month, int year, int hour, int minute, int second)
        {
            CheckIfEntityExists();
            AltNative.Player.Player_SetDateTime(NativePointer, day, month, year, hour, minute, second);
        }

        public void SetWeather(uint weather)
        {
            CheckIfEntityExists();
            AltNative.Player.Player_SetWeather(NativePointer, weather);
        }

        public void GiveWeapon(uint weapon, int ammo, bool selectWeapon)
        {
            CheckIfEntityExists();
            AltNative.Player.Player_GiveWeapon(NativePointer, weapon, ammo, selectWeapon);
        }

        public void RemoveWeapon(uint weapon)
        {
            CheckIfEntityExists();
            AltNative.Player.Player_RemoveWeapon(NativePointer, weapon);
        }

        public void RemoveAllWeapons()
        {
            CheckIfEntityExists();
            AltNative.Player.Player_RemoveAllWeapons(NativePointer);
        }

        public void Kick(string reason)
        {
            CheckIfEntityExists();
            var reasonPtr = AltNative.StringUtils.StringToHGlobalUtf8(reason);
            AltNative.Player.Player_Kick(NativePointer, reasonPtr);
            Marshal.FreeHGlobal(reasonPtr);
        }

        public void Emit(string eventName, params object[] args)
        {
            CheckIfEntityExists();
            Alt.Server.TriggerClientEvent(this, eventName, args);
        }

        public ReadOnlyPlayer Copy()
        {
            CheckIfEntityExists();
            var readOnlyPlayer = ReadOnlyPlayer.Empty;
            AltNative.Player.Player_Copy(NativePointer, ref readOnlyPlayer);
            return readOnlyPlayer;
        }
    }
}

File: runtime/src/altv-c-api/player.cpp
#include "player.h"

// Entity

uint16_t Player_GetID(alt::IPlayer* player) {
    return player->GetID();
}

uint32_t Player_GetModel(alt::IPlayer* player) {
    return player->GetModel();
}

void Player_SetModel(alt::IPlayer* player, uint32_t model) {
    player->SetModel(model);
}

void Player_GetPosition(alt::IPlayer* player, position_t &position) {
    auto playerPosition = player->GetPosition();
    position.x = playerPosition.x;
    position.y = playerPosition.y;
    position.z = playerPosition.z;
}

void Player_SetPosition(alt::IPlayer* player, alt::Position pos) {
    player->SetPosition(pos);
}

void Player_GetRotation(alt::IPlayer* player, rotation_t &rotation) {
    auto playerRotation = player->GetRotation();
    rotation.roll = playerRotation.roll;
    rotation.pitch = playerRotation.pitch;
    rotation.yaw = playerRotation.yaw;
}

void Player_SetRotation(alt::IPlayer* player, alt::Rotation rot) {
    player->SetRotation(rot);
}

int16_t Player_GetDimension(alt::IPlayer* player) {
    return player->GetDimension();
}

void Player_SetDimension(alt::IPlayer* player, uint16_t dimension) {
    player->SetDimension(dimension);
}

void Player_GetMetaData(alt::IPlayer* player, const char* key, alt::MValue &val) {
    val = player->GetMetaData(key);
}

void Player_SetMetaData(alt::IPlayer* player, const char* key, alt::MValue* val) {
    player->SetMetaData(key, *val);
}

void Player_GetSyncedMetaData(alt::IPlayer* player, const char* key, alt::MValue &val) {
    val = player->GetSyncedMetaData(key);
}

void Player_SetSyncedMetaData(alt::IPlayer* player, const char* key, alt::MValue* val) {
    player->SetSyncedMetaData(key, *val);
}

// Player

bool Player_IsConnected(alt::IPlayer* player) {
    return player->IsConnected();
}

void Player_Spawn(alt::IPlayer* player, alt::Position pos, uint32_t delayMs) {
    player->Spawn(pos, delayMs);
}

void Player_Despawn(alt::IPlayer* player) {
    player->Despawn();
}

void Player_GetName(alt::IPlayer* player, const char*&name) {
    name = player->GetName().CStr();
}

void Player_SetName(alt::IPlayer* player, const char* name) {
    player->SetName(name);
}

uint16_t Player_GetHealth(alt::IPlayer* player) {
    return player->GetHealth();
}

void Player_SetHealth(alt::IPlayer* player, uint16_t health) {
    player->SetHealth(health);
}

void Player_SetDateTime(alt::IPlayer* player, int day, int month, int year, int hour, int minute, int second) {
    player->SetDateTime(day, month, year, hour, minute, second);
}

void Player_SetWeather(alt::IPlayer* player, uint32_t weather) {
    player->SetWeather(weather);
}

void Player_GiveWeapon(alt::IPlayer* player, uint32_t weapon, int32_t ammo, bool selectWeapon) {
    player->GiveWeapon(weapon, ammo, selectWeapon);
}

void Player_RemoveWeapon(alt::IPlayer* player, uint32_t weapon) {
    player->RemoveWeapon(weapon);
}

void Player_RemoveAllWeapons(alt::IPlayer* player) {
    player->RemoveAllWeapons();
}

bool Player_IsDead(alt::IPlayer* player) {
    return player->IsDead();
}

bool Player_IsJumping(alt::IPlayer* player) {
    return player->IsJumping();
}

bool Player_IsInRagdoll(alt::IPlayer* player) {
    return player->IsInRagdoll();
}

bool Player_IsAiming(alt::IPlayer* player) {
    return player->IsAiming();
}

bool Player_IsShooting(alt::IPlayer* player) {
    return player->IsShooting();
}

bool Player_IsReloading(alt::IPlayer* player) {
    return player->IsReloading();
}

uint16_t Player_GetArmor(alt::IPlayer* player) {
    return player->GetArmor();
}

void Player_SetArmor(alt::IPlayer* player, uint16_t armor) {
    player->SetArmor(armor);
}

float Player_GetMoveSpeed(alt::IPlayer* player) {
    return player->GetMoveSpeed();
}

uint32_t Player_GetWeapon(alt::IPlayer* player) {
    return player->GetWeapon();
}

uint16_t Player_GetAmmo(alt::IPlayer* player) {
    return player->GetAmmo();
}

void Player_GetAimPos(alt::IPlayer* player, position_t &aimPosition) {
    auto playerAimPosition = player->GetAimPos();
    aimPosition.x = playerAimPosition.x;
    aimPosition.y = playerAimPosition.y;
    aimPosition.z = playerAimPosition.z;
}

void Player_GetHeadRotation(alt::IPlayer* player, rotation_t &headRotation) {
    auto playerHeadRotation = player->GetHeadRotation();
    headRotation.roll = playerHeadRotation.roll;
    headRotation.pitch = playerHeadRotation.pitch;
    headRotation.yaw = playerHeadRotation.yaw;
}

bool Player_IsInVehicle(alt::IPlayer* player) {
    return player->IsInVehicle();
}

alt::IVehicle* Player_GetVehicle(alt::IPlayer* player) {
    return player->GetVehicle();
}

uint8_t Player_GetSeat(alt::IPlayer* player) {
    return player->GetSeat();
}

void Player_Kick(alt::IPlayer* player, const char* reason) {
    player->Kick(reason);
}

uint32_t Player_GetPing(alt::IPlayer* player) {
    return player->GetPing();
}

void Player_Copy(alt::IPlayer* player, player_struct_t* player_struct) {
    player_struct->id = player->GetID();
    auto position = player->GetPosition();
    player_struct->position.x = position.x;
    player_struct->position.y = position.y;
    player_struct->position.z = position.z;
    auto rotation = player->GetRotation();
    player_struct->rotation.roll = rotation.roll;
    player_struct->rotation.pitch = rotation.pitch;
    player_struct->rotation.yaw = rotation.yaw;
    player_struct->dimension = player->GetDimension();
    player_struct->ping = player->GetPing();
    player_struct->model = player->GetModel();
    player_struct->seat = player->GetSeat();
    auto aimPosition = player->GetAimPos();
    player_struct->aim_position.x = aimPosition.x;
    player_struct->aim_position.y = aimPosition.y;
    player_struct->aim_position.z = aimPosition.z;
    auto headRotation = player->GetHeadRotation();
    player_struct->head_rotation.roll = headRotation.roll;
    player_struct->head_rotation.pitch = headRotation.pitch;
    player_struct->head_rotation.yaw = headRotation.yaw;
    player_struct->armor = player->GetArmor();
    player_struct->move_speed = player->GetMoveSpeed();
    auto name = player->GetName();
    // Free in c# after async method ends
    auto copiedName = new char[name.GetSize() + 1];
    memcpy(copiedName, name.GetData(), name.GetSize());
    copiedName[name.GetSize()] = '\0';
    player_struct->name = copiedName;
    player_struct->health = player->GetHealth();
    player_struct->is_in_ragdoll = player->IsInRagdoll();
    player_struct->is_dead = player->IsDead();
    player_struct->is_shooting = player->IsShooting();
    player_struct->is_aiming = player->IsAiming();
    player_struct->is_in_vehicle = player->IsInVehicle();
    player_struct->is_jumping = player->IsJumping();
    player_struct->is_reloading = player->IsReloading();
    player_struct->is_connected = player->IsConnected();
    player_struct->vehicle = player->GetVehicle();
}

void Player_Copy_Dispose(player_struct_t* player_struct) {
    delete[] player_struct->name;
}